File: leetcode/1119/main.js
/**
 * @param {string}
 * @return {string}
 */
// Runtime: 76 ms, faster than 73.24% of JavaScript online submissions for Remove Vowels from a String.
// Memory Usage: 38.6 MB, less than 47.49% of JavaScript online submissions for Remove Vowels from a String.
var removeVowels = function(s) {
    
    // let result = ""
    let result = []
    var p = 0
    for (var i = 0; i < s.length; i++) {
        switch (s[i]) {
            case 'a':
            case 'e':
            case 'i':
            case 'o':
            case 'u':
                break;
            default:
                result[p] = s[i];
                p++;
        }
    }

    result = result.join("");
    
    return result
};

/**
 * @param {string}
 * @return {string}
 */
// Runtime: 80 ms, faster than 43.26% of JavaScript online submissions for Remove Vowels from a String.
// Memory Usage: 38.8 MB, less than 19.92% of JavaScript online submissions for Remove Vowels from a String.
var removeVowels = function(s) {
    
    let result = ""
    for (var i = 0; i < s.length; i++) {
        switch (s[i]) {
            case 'a':
            case 'e':
            case 'i':
            case 'o':
            case 'u':
                break;
            default:
                result += s[i]
        }
    }
    
    return result
};

File: leetcode/1119/main.go
package main

import (
	"fmt"
	"strings"
)

func main() {

	s := "leetcodeisacommunityforcoders"
	e := "ltcdscmmntyfrcdrs"
	r := removeVowelsSB(s)

	fmt.Printf("result matches expectation? [%v]\n", e == r)

	s = "aeiou"
	e = ""
	r = removeVowelsSB(s)

	fmt.Printf("result matches expectation? [%v]\n", e == r)
}

var vowels = map[rune]bool{
	'a': true,
	'e': true,
	'i': true,
	'o': true,
	'u': true,
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Remove Vowels from a String.
// Memory Usage: 2.9 MB, less than 32.14% of Go online submissions for Remove Vowels from a String.
func removeVowels(s string) string {

	var r string
	// o(n)
	for _, c := range s {
		// O(1)
		if _, vowel := vowels[c]; !vowel {
			// ? use a string builder
			r += string(c)
		}
	}

	return r
}

/*
Runtime: 0 ms, faster than 100.00% of Go online submissions for Remove Vowels from a String.
Memory Usage: 2.1 MB, less than 50.00% of Go online submissions for Remove Vowels from a String.
*/
func removeVowelsSB(s string) string {

	var r strings.Builder
	// o(n)
	for _, c := range s {
		// O(1)
		if _, ok := vowels[c]; !ok {
			// use a string builder
			fmt.Fprintf(&r, "%s", string(c))
		}
	}

	return r.String()
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Remove Vowels from a String.
// Memory Usage: 2.1 MB, less than 42.86% of Go online submissions for Remove Vowels from a String.
func removeVowelsSwitch(s string) string {

	var r strings.Builder
	// o(n)
	for _, c := range s {
		// O(1)
		switch c {
		case 'a', 'e', 'i', 'o', 'u':
			break
		default:
			// use a string builder
			fmt.Fprintf(&r, "%s", string(c))
		}
	}

	return r.String()
}
